service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,240 @@
+        }
+    }
+
+    @Override
+    public List<GiftCertificate> getGiftCertificates(CertificateRequestBody requestBody) throws ServiceException {
+        if (requestBody == null) {
+            return geAllCertificates();
+        } else {
+            return getGiftCertificatesByRequestBody(requestBody);
+        }
+    }
+
+    private List<GiftCertificate> getGiftCertificatesByRequestBody(CertificateRequestBody requestBody)
+            throws ServiceException {
+        if (requestBody.getContent() != null) {
+            return getGiftCertificatesByContent(requestBody.getContent());
+        }
+        if (requestBody.getSortType() != null && requestBody.getSortBy() != null) {
+            return getSortedCertificates(requestBody.getSortType(), requestBody.getSortBy());
+        }
+        if (requestBody.getTagName() != null) {
+            return getGiftCertificateByTagName(requestBody.getTagName());
+        }
+
+        throw new ServiceException("Error: request input body is empty", ErrorCodeEnum.INVALID_SORT_INPUT);
+    }
+
+    private List<GiftCertificate> getSortedCertificates(SortType sortType, SortParameter sortBy)
+            throws ServiceException {
+        switch (sortBy) {
+            case DATE : return getAllGiftCertificatesSortedByDate(isAscending(sortType));
+            case NAME : return getAllGiftCertificatesSortedByName(isAscending(sortType));
+            default: throw new NoSuchElementException("Not found sort criteria");
+        }
+    }
+
+    private boolean isAscending(SortType sort) {
+        return sort == SortType.ASC;
+    }
+
+    @Override
+    @Transactional
+    public GiftCertificate addGiftCertificate(GiftCertificate giftCertificate) throws ServiceException {
+        certificateValidator.validateCertificate(giftCertificate);
+        try {
+            giftCertificate.setCreateDate(ZonedDateTime.ofInstant(Instant.now(), ZoneOffset.UTC));
+            giftCertificate.setLastUpdateDate(giftCertificate.getCreateDate());
+
+            int id = giftCertificateDAO.addGiftCertificate(giftCertificate);
+            giftCertificate.setId(id);
+
+            addNewTagsToCertificate(giftCertificate);
+
+            return giftCertificate;
+        } catch (DataAccessException e) {
+            LOGGER.error("Following exception was thrown in addGiftCertificate(): " + e.getMessage());
+            throw new ServiceException("Failed to add certificate: certificate already exist",
+                    ErrorCodeEnum.FAILED_TO_ADD_CERTIFICATE);
+        } catch (PersistenceException e) {
+            LOGGER.error("Following exception was thrown: " + e.getMessage());
+            throw new ServiceException(e.getMessage(), ErrorCodeEnum.FAILED_TO_ADD_CERTIFICATE);
+        }
+    }
+
+    @Override
+    @Transactional
+    public void deleteGiftCertificate(int id) throws ServiceException {
+        GiftCertificate giftCertificate = getGiftCertificate(id);
```

Comment:
First check if exist, if not throw exception

Reply:
Done